src/vs/workbench/contrib/terminal/browser/xterm/decorationAddon.ts

```diff
@@ -221,7 +221,22 @@ export class DecorationAddon extends Disposable implements ITerminalAddon, IDeco
 			this.registerCommandDecoration(command);
 		}
 		commandDetectionListeners.push(capability.onCommandFinished(command => {
-			this.registerCommandDecoration(command);
+			const buffer = this._terminal?.buffer?.active;
+			const marker = command.marker;
+
+			// Only apply cursor check for commands with exit codes (completed commands)
+			// Skip the check for incomplete commands (no exitCode) to allow placeholders
+			if (command.exitCode !== undefined && buffer && marker) {
+				// Only register if cursor is at or below the command start marker
+				// This prevents misplaced decorations when watch commands clear previous output
+				if (buffer.baseY + buffer.cursorY >= marker.line) {
+					this.registerCommandDecoration(command);
+				}
+			} else {
+				// No exit code means incomplete command - always register placeholder
+				this.registerCommandDecoration(command);
+			}
```

Comment:
Can it be structured more like this?

```ts
// Explain the edge case we're handling
const shouldRegisterDecoration = (
  clause1
  // comment if needs explanation
  clause2
);
if (shouldRegisterDecoration) {
  this.registerCommandDecoration(command);
}
```

Example:

https://github.com/microsoft/vscode/blob/bd5b0024c264399c45cbacb21998386056cdc936/src/vs/workbench/contrib/terminal/browser/terminalService.ts#L648-L654